crystal: 1.11 -> 1.14

will opened the pull request:
## Current Description of changes

Release notes: https://crystal-lang.org/2024/10/09/1.14.0-released/

## Old Description of changes

Release notes: https://crystal-lang.org/2024/04/09/1.12.0-released

From the notes: 

> Our own LLVM extension library libllvm_ext is no longer required when linking LLVM >= 18 because all APIs are now directly exposed in libllvm directly. The Makefile automatically skips building libllvm_ext.o on LLVM 18 and above.

So I've updated the llvm package for this version to llvm 18. From talking with people on the Crystal core team, there is very small chance that the 16-bit alignment for 128-bit integers in llvm18 can cause problems on x86 systems (not aarch64 since they already were using 16-bit alignment), and if so we should go down to llvm 17.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)

Result of `nixpkgs-review` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>rtfm</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>collision</li>
    <li>crystalline</li>
    <li>invidious</li>
  </ul>
</details>
<details>
  <summary>18 packages built:</summary>
  <ul>
    <li>ameba</li>
    <li>amqpcat</li>
    <li>blahaj</li>
    <li>crystal</li>
    <li>crystal.bin</li>
    <li>crystal.lib</li>
    <li>crystal2nix</li>
    <li>gi-crystal</li>
    <li>icr</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>mint</li>
    <li>oq</li>
    <li>scry</li>
    <li>shards (shards_0_17)</li>
    <li>thicket</li>
    <li>tijolo</li>
    <li>tmuxPlugins.fingers</li>
  </ul>
</details>



- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).



<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

will committed:
crystal: 1.11 -> 1.12

cole-h commented:
You'll want to rebase on the latest `master` -- there was an issue that broke ofborg's ability to evaluate Nixpkgs that has since been resolved. Sorry for the inconvenience!

will committed:
crystalline: use llvm18 to match crystal 1.12

will committed:
invidious: add patch to fix compile error

will committed:
crystal: add missing versions to all-packages.nix

will committed:
collision: mark as broken

will commented:
I've also now fixed crystalline by making sure to pick the same version of llvm. And I've patched a type error in invidious that I would assume they'll get to eventually in a new version and nixpkgs can drop that small patch when it updates to a newer version of invidious. 

The `collision` package, however, I can't get to work. There is an error generated while running its Makefile that is from `gi-crystal`. No combination of pinning back to previous versions of crystal for either `collision` and/or `gi-crystal` seems to fix it, so I've marked it as broken for now. 

Result of `nixpkgs-review` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>collision</li>
    <li>rtfm</li>
  </ul>
</details>
<details>
  <summary>23 packages built:</summary>
  <ul>
    <li>ameba</li>
    <li>amqpcat</li>
    <li>blahaj</li>
    <li>crystal (crystal_1_12)</li>
    <li>crystal.bin (crystal_1_12.bin)</li>
    <li>crystal.lib (crystal_1_12.lib)</li>
    <li>crystal2nix</li>
    <li>crystal_1_11</li>
    <li>crystal_1_11.bin</li>
    <li>crystal_1_11.lib</li>
    <li>crystalline</li>
    <li>gi-crystal</li>
    <li>icr</li>
    <li>invidious</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>mint</li>
    <li>oq</li>
    <li>scry</li>
    <li>shards (shards_0_17)</li>
    <li>thicket</li>
    <li>tijolo</li>
    <li>tmuxPlugins.fingers</li>
  </ul>
</details>

peterhoeg commented:
We don't share homebrew's problems, but I just want to bring attention to https://github.com/crystal-lang/crystal/pull/14318 to see if anyone has any issues that this might solve.

will commented:
@peterhoeg I saw that patch too, and while I'm not 100% sure, I think think the long stdenv wrapper scripts around the linker and whatnot take care of the problems that patch is trying to address

will committed:
crystal: 1.12.0 -> 1.12.1

peterhoeg commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15546,12 +15546,14 @@ with pkgs;
 
   crystalline = callPackage ../development/tools/language-servers/crystalline {
-    llvmPackages = llvmPackages_15;
+    llvmPackages = llvmPackages_18;
```
Instead of manually defining it here, why not expose the used llvm through `passthru` on `crystal` and use that? This way they will always be in sync.

will replied:
Good idea! I pushed up a commit to do that just now

will committed:
crystal: add llvmPackages passthru and use it for crystalline

sund3RRR commented:
I can't compile anything on EndeavourOS regardless of crystal version
```bash
nix-build -A tijolo
this derivation will be built:
  /nix/store/pwlcp6r8x2lwg6d0xq0aficw319nj3bw-tijolo-0.7.4.drv
building '/nix/store/pwlcp6r8x2lwg6d0xq0aficw319nj3bw-tijolo-0.7.4.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/qiv3kriqmw0i5kpqrgaqsb1nbrr6a1bx-source
source root is source
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
Running phase: buildPhase
shards build --release -s  -Dpreview_mt --ignore-crystal-version
Dependencies are satisfied
Building: tijolo
Parse:                             00:00:00.000122163 (   1.07MB)
Error target tijolo failed to compile:
impure path `/homeless-shelter/.cache/crystal/build-source-lib-version_from_shard-src-extract_version.cr/macro_run' used in link
collect2: error: ld returned 1 exit status
Error: execution of command failed with exit status 1: gcc "${@}" -o /homeless-shelter/.cache/crystal/build-source-lib-version_from_shard-src-extract_version.cr/macro_run  -rdynamic -L/nix/store/vyz5cpkg7rli3236cy30dald1yj9mdch-boehm-gc-8.2.4/lib -L/nix/store/5sqdrc4jpr4vjiiqycyw8q4v3zchpdka-pcre2-10.43/lib -L/nix/store/jlqg0qc06nas4qcpqn8w9nnkgsvrxjxp-libevent-2.1.12/lib -L/nix/store/pqjjikxz0xb7g79gj1h75qbgxg2xkbkz-libyaml-0.2.5/lib -L/nix/store/zph9xw0drmq3rl2ik5slg0n2frw9lw5m-zlib-1.3.1/lib -L/nix/store/ii7r1shwrbl9rcv6nn6mfvih5qcb742p-libxml2-2.12.6/lib -L/nix/store/p25ghy7y53lyc834xnw5mrhfq096wa4x-openssl-3.0.13/lib -L/nix/store/pqjjikxz0xb7g79gj1h75qbgxg2xkbkz-libyaml-0.2.5/lib -lyaml -lpcre2-8 -L/nix/store/vyz5cpkg7rli3236cy30dald1yj9mdch-boehm-gc-8.2.4/lib -lgc -lpthread -ldl -lpthread -L/nix/store/jlqg0qc06nas4qcpqn8w9nnkgsvrxjxp-libevent-2.1.12/lib -levent_pthreads -levent -L/nix/store/jlqg0qc06nas4qcpqn8w9nnkgsvrxjxp-libevent-2.1.12/lib -levent -lrt -lpthread -ldl

make: *** [Makefile:5: all] Error 1
error: builder for '/nix/store/pwlcp6r8x2lwg6d0xq0aficw319nj3bw-tijolo-0.7.4.drv' failed with exit code 2
```

Probably it is a non-nixos issue, but I think this shouldn't happen.

straight-shoota commented on pkgs/servers/invidious/default.nix:
```diff
@@ -38,6 +38,13 @@ crystal.buildCrystalPackage rec {
       assetCommitTemplate = ''{{ "#{`git rev-list HEAD --max-count=1 --abbrev-commit -- assets`.strip}" }}'';
     in
     ''
+      # bugfix for crystal_1_12 Time::Span, can be removed when fixed upstream
+      substituteInPlace src/invidious/yt_backend/proxy.cr \
+        --replace-fail "opts[:dns_timeout] = @dns_timeout" "opts[:dns_timeout] = @dns_timeout.try(&.to_f)" \
+        --replace-fail "opts[:connect_timeout] = @connect_timeout" "opts[:connect_timeout] = @connect_timeout.try(&.to_f)" \
+        --replace-fail "opts[:read_timeout] = @read_timeout" "opts[:read_timeout] = @read_timeout.try(&.to_f)" \
```
This is fixed in the latest release https://github.com/iv-org/invidious/releases/tag/v2.20240427 (cf. https://github.com/iv-org/invidious/pull/4570). The file `src/invidious/yt_backend/proxy.cr` was removed entirely upstream.

will committed:
crystal: 1.12.1 -> 1.14.0

will committed:
tijolo: mark as broken

While updating crystal for 1.14 this project breaks with

    In src/code_language.cr:46:18

     46 | mime = Gio.content_type_guess(file.to_s, contents[0, contents_size])
                     ^-----------------
    Error: wrong number of arguments for 'Gio.content_type_guess' (given 2, expected 4)

    Overloads are:
     - Gio.content_type_guess(filename : ::String, data : Pointer(UInt8), data_size : UInt64, result_uncertain : Pointer(Bool))

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4669

will commented:
ping @david50407 @manveru @peterhoeg @donovanglover

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/ysk-guide-on-getting-your-pr-merged/54513/2

donovanglover commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 303173`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>tijolo</li>
  </ul>
</details>
<details>
  <summary>:x: 21 packages failed to build:</summary>
  <ul>
    <li>ameba</li>
    <li>amqpcat</li>
    <li>blahaj</li>
    <li>collision</li>
    <li>crystal (crystal_1_14)</li>
    <li>crystal.bin (crystal_1_14.bin)</li>
    <li>crystal.lib (crystal_1_14.lib)</li>
    <li>crystal2nix</li>
    <li>crystalline</li>
    <li>gi-crystal</li>
    <li>icr</li>
    <li>invidious</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>mint</li>
    <li>oq</li>
    <li>rtfm</li>
    <li>scry</li>
    <li>shards (shards_0_17)</li>
    <li>thicket</li>
    <li>tmuxPlugins.fingers</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 6 packages built:</summary>
  <ul>
    <li>crystal_1_11</li>
    <li>crystal_1_11.bin</li>
    <li>crystal_1_11.lib</li>
    <li>crystal_1_12</li>
    <li>crystal_1_12.bin</li>
    <li>crystal_1_12.lib</li>
  </ul>
</details>

donovanglover commented:
Seems to fail during `checkPhase` from the following:
```crystal
In lib/markd/src/markd/renderers/html_renderer.cr:238:7

 238 | {% if Crystal::VERSION < "1.2.0" %}
       ^
Warning: expanding macro


There was a problem expanding macro 'macro_140735611328032'

Called macro defined in lib/markd/src/markd/renderers/html_renderer.cr:238:7

 238 | {% if Crystal::VERSION < "1.2.0" %}

Which expanded to:

   1 | 
 > 2 |         title = URI.encode(node.first_child.text)
   3 |         @output_io << %(<a id="anchor-) << title << %(" class="anchor" href="#anchor-) << title << %("></a>)
   4 |       
Warning: Deprecated URI.encode. Use `.encode_path` instead.

A total of 1 warnings were found.
```
Could add `doCheck = false;` since it'd also reduce the build time a bit.

straight-shoota commented:
Warnings should really not break a build. They're just warnings.

donovanglover commented:
Seems to be caused by `make: *** [Makefile:110: compiler_spec] Illegal instruction (core dumped)` later in the log

will commented:
@donovanglover thanks so much for taking the time to check this!

I just got ahold of an x86 linux VM and am seeing if I can figure out what's up

donovanglover commented:
FWIW I'm okay with disabling `checkPhase` if Crystal still works. Not sure how hard it is to fix the tests.

will committed:
crystal: set doCheck to false for 1.14.0

donovanglover reviewed (approved):
LGTM

will commented:
Thanks @donovanglover for all the time and help with this!
